std_execution.bs

```diff
@@ -3032,6 +3174,48 @@ are all well-formed.
 
     3. Returns an operation state `op_state` that contains `op_state2`. When `execution::start(op_state)` is called, calls `execution::start(op_state2)`.
 
+#### `execution::done_as_optional` <b>[execution.senders.adaptors.done_as_optional]</b> #### {#spec-execution.senders.adaptors.done_as_optional}
+
+1. `execution::done_as_optional` is used to handle a done signal by mapping it into the value channel as an empty optional. The value channel is also converted into an optional. The result is a sender that never completes with done, reporting cancellation by completing with an empty optional.
+
+2. The name `execution::done_as_optional` denotes a customization point object. For some subexpression `s.`, let `S` be `decltype((s))`. If the type `S` does not satisfy <code><i>single-typed-sender</i></code>, `execution::done_as_optional(s)` is ill-formed. Otherwise, the expression `execution::done_as_optional(s)` is expression-equivalent to:
+
+    <pre highlight=c++>
+        execution::let_done(
+          execution::then(s, [](auto&& t) {
+              return optional&lt;<i>single-sender-value-type</i>&lt;S>>{
+                static_cast&lt;decltype(t)>(t)
+              };
+            }
+          ),
+          [] {
+            return execution::just(optional&lt;<i>single-sender-value-type</i>&lt;S>>{});
+          }
+        )
+    </pre>
+
+#### `execution::done_as_exception` <b>[execution.senders.adaptors.done_as_exception]</b> #### {#spec-execution.senders.adaptors.done_as_exception}
+
+1. `execution::done_as_exception` is used to handle a done signal by mapping it into the error channel as an `exception_ptr` that refers to a custom exception type. The result is a sender that never completes with done, reporting cancellation by completing with an error.
+
+2. The template <code><i>into-variant-type</i></code> is used to compute the type sent by a sender returned from `execution::into_variant`.
+
+    <pre highlight=c++>
+        template&lt;default_initializable Exception, sender S>
+          <i>see-below</i> done_as_exception(S && s);
+    </pre>
+
+3. <i>Effects:</i> Equivalent to:
+
+    <pre highlight=c++>
+        return execution::let_done(
```

Comment:
The implementation in terms of `let_done()`, while functional is not optimal as it forces the operation-state to store a copy of the value. Whereas if it were hand-implemented then we could just have the implementation of `set_done` just call `set_error(std::move(self.r), Error{})` and not have to store the error.

Reply:
We can revisit that in R3.